examples/perf/ipc.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/python
+# IPC - Instructions Per Cycles using Perf Events and
+# uprobes
+# 24-Apr-2020	Saleem Ahmad	Created this.
+
+from __future__ import print_function
+from bcc import BPF, USDT, Perf
+from bcc.utils import printb
+from time import sleep
+import sys
+from optparse import OptionParser
+
+# load BPF program
+code="""
+#include <uapi/linux/ptrace.h>
+
+struct perf_delta {
+    u64 clk_delta;
+    u64 inst_delta;
+    u64 time_delta;
+};
+
+// Perf Arrays to read counter values for open
+// perf events.
+BPF_PERF_ARRAY(clk, 16);
+BPF_PERF_ARRAY(inst, 16);
+
+// Perf Output
+BPF_PERF_OUTPUT(output);
+
+// Per Cpu Data to store start values
+BPF_PERCPU_ARRAY(data, u64);
+
+void trace_start(struct pt_regs *ctx) {
+    u32 clk_k = 0;
+    u32 inst_k = 1;
+    u32 time = 2;
+
+    int cpu = bpf_get_smp_processor_id();
+    u64 clk_start = clk.perf_read(cpu);
+    u64 inst_start = inst.perf_read(cpu);
```

Comment:
perf_read may return negative value for errors. In this particular case, if "cpu" id is more than the perf array 'clk'/'inst' size, you will get an enormous large number of clk_start/inst_start and things will go wrong.

Most people do not check this as they know their max entry in the perf array greater than the number of cpus in the system, but a comment will be good. See above.

Reply:
Added comment for same.